src/compiler/emitter.ts

```diff
@@ -4866,20 +4867,27 @@ export function createPrinter(printerOptions: PrinterOptions = {}, handlers: Pri
     }
 
     function emitEmbeddedStatement(parent: Node, node: Statement) {
-        if (isBlock(node) || getEmitFlags(parent) & EmitFlags.SingleLine) {
+        const lines = getLeadingLineTerminatorCount(parent, node, ListFormat.None);

```

Comment:
I use getLeadingLineTerminatorCount to determine if there is any lines/space between the first node and its parent node under the condition that `preserveSourceNewlines` is `true`.